### PR TITLE
Add quota workflow for manifest

### PR DESCRIPTION
**Pull request by wy65701436**

### CODE REVIEW

File: src/core/middlewares/registryproxy/handler.go [hunk 1]
```diff
@@ -117,6 +110,96 @@ func singleJoiningSlash(a, b string) string {
...
+
+	} else if res.StatusCode >= 400 || res.StatusCode <= 511 {
+		success := subtractResources(mf)
+		if !success {
+			return errors.New("Error to release resource booked for the manifest")
+		}
+	}
+
+	defer func() {
+		_, err := mf.TagLock.Free()
```
Comment (heww) [inline src/core/middlewares/registryproxy/handler.go]: Suggest to move this `defer func` to top.

File: src/core/middlewares/util/util.go [hunk 2]
```diff
@@ -48,6 +53,30 @@ type ImageInfo struct {
 	Digest      string
 }
 
+// MfInfo ...
+type MfInfo struct {
```
Comment (heww) [inline src/core/middlewares/util/util.go]: `ManifestInfo` will be better?

File: src/core/middlewares/registryproxy/handler.go (hunk 1 above)
Comment (wy65701436) [reply src/core/middlewares/registryproxy/handler.go]: ok

File: src/core/middlewares/regquota/handler.go [hunk 3]
```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
...
-					http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to decode manifest body %v", err)), http.StatusInternalServerError)
-					return
-				}
-				req.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+		if mediaType == schema1.MediaTypeManifest ||
+			mediaType == schema1.MediaTypeSignedManifest ||
+			mediaType == schema2.MediaTypeManifest {
+
+			tagLock, err := rqh.tryLockTag()
+			if err != nil {
```
Comment (ywk253100) [inline src/core/middlewares/regquota/handler.go]: release the lock?

File: src/core/middlewares/regquota/handler.go [hunk 4]
```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
...
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to Unmarshal Manifest %v", err)), http.StatusInternalServerError)
+				return
+			}
+			rqh.mfInfo.Refrerence = manifest.References()
+			rqh.mfInfo.Digest = desc.Digest.String()
+			rqh.mfInfo.Size = desc.Size
+
+			projectID, err := rqh.getProjectID(strings.Split(repository, "/")[0])
+			if err != nil {
+				log.Warningf("Error occurred when to get project ID %v", err)
```
Comment (ywk253100) [inline src/core/middlewares/regquota/handler.go]: release lock

File: src/core/middlewares/regquota/handler.go [hunk 5]
```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
...
+				return
+			}
+			rqh.mfInfo.ProjectID = projectID
+
+			exist, af, err := rqh.imageExist()
```
Comment (ywk253100) [inline src/core/middlewares/regquota/handler.go]: Should split the logic of handling storage quota and number quota here

File: src/core/middlewares/regquota/handler.go [hunk 6]
```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
...
-				mfDigest = desc.Digest.String()
-				mfSize = desc.Size
-				log.Infof("manifest digest... %s", mfDigest)
-				log.Infof("manifest size... %v", mfSize)
 			}
+			quotaRes, err := rqh.tryRequireQuota()
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Errorf("Cannot get quota for the manifest %v", err)
+				http.Error(rw, util.MarshalError("StatusNotAcceptable", fmt.Sprintf("Cannot get quota for the manifest %v", err)), http.StatusNotAcceptable)
```
Comment (ywk253100) [inline src/core/middlewares/regquota/handler.go]: How can we determine whether the error is an `internal server error `or `quota not meet the requirement error`?

File: src/core/middlewares/regquota/handler.go [hunk 7]
```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
...
+		quotaRes = quota.ResourceList{
+			quota.ResourceStorage: rqh.mfInfo.Size,
+			quota.ResourceCount:   1,
+		}
+	} else if rqh.mfInfo.DigestChanged {
+		quotaRes = quota.ResourceList{
+			quota.ResourceStorage: rqh.mfInfo.Size,
+			quota.ResourceCount:   0,
+		}
+	}
```
Comment (ywk253100) [inline src/core/middlewares/regquota/handler.go]: For the other cases(not included in the `if else block`), we shouldn't try to change the quota at all.

File: src/core/middlewares/regquota/handler.go [hunk 8]
```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
...
+
+			exist, af, err := rqh.imageExist()
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Warningf("Error occurred when to check Manifest existence %v", err)
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to check Manifest existence %v", err)), http.StatusInternalServerError)
+				return
+			}
+			rqh.mfInfo.Exist = exist
+			if exist {
```
Comment (ywk253100) [inline src/core/middlewares/regquota/handler.go]: If it already exists, can we release the lock here?

File: src/core/middlewares/regquota/handler.go (hunk 3 above)
Comment (wy65701436) [reply src/core/middlewares/regquota/handler.go]: cannot release here

File: src/core/middlewares/regquota/handler.go (hunk 4 above)
Comment (wy65701436) [reply src/core/middlewares/regquota/handler.go]: ok

File: src/core/middlewares/regquota/handler.go (hunk 6 above)
Comment (wy65701436) [reply src/core/middlewares/regquota/handler.go]: define an new error

File: src/core/middlewares/regquota/handler.go (hunk 5 above)
Comment (wy65701436) [reply src/core/middlewares/regquota/handler.go]: yes, move the logic on size to the blob handler.

**Pull request merged**